domain/schema/model.go

```diff
@@ -401,23 +402,246 @@ ON charm_storage (charm_uuid);
 `)
 }
 
+func constraintsSchema() schema.Patch {
+	return schema.MakePatch(`
+-- constraint is a reserved word in sqlite, so we use juju_constraint instead.
+CREATE TABLE juju_constraint (
+    uuid                TEXT PRIMARY KEY,
+    arch                TEXT,
+    cpu_cores           INT,
+    cpu_power           INT,
+    mem                 INT,
+    root_disk           INT,
+    root_disk_source    TEXT,
+    instance_role       TEXT,
+    instance_type       TEXT,
+    container_type_id   INT,
+    virt_type           TEXT,
+    allocate_public_ip  BOOLEAN,
+    image_id            TEXT,
+    CONSTRAINT          fk_constraint_container_type
+        FOREIGN KEY     (container_type_id)
+        REFERENCES      container_type(id)
+);
+
+CREATE TABLE constraint_tag (
+    id                  INT PRIMARY KEY,
+    constraint_uuid     TEXT NOT NULL,
+    tag                 TEXT NOT NULL,
+    CONSTRAINT          fk_constraint_tag_constraint
+        FOREIGN KEY     (constraint_uuid)
+        REFERENCES      juju_constraint(uuid)
+);
+
+CREATE TABLE constraint_space (
+    id                  INT PRIMARY KEY,
+    constraint_uuid     TEXT NOT NULL,
+    space               TEXT NOT NULL,
```

Comment:
Should we be FK'ing to the constrained resource (in this case `space` to `space(name)`)?

Reply:
agreed